Improve MediaKeySession.generateRequest() details (parameters, exceptions) (#39356)
mdn/content: files/en-us/web/api/mediakeysession/generaterequest/index.md

{{APIRef("Encrypted Media Extensions")}}{{SecureContext_Header}}

> The `generateRequest()` method of the {{domxref('MediaKeySession')}} interface returns a {{jsxref('Promise')}} after generating a licence request based on initialization data.

Reviewer: ```suggestion
The `generateRequest()` method of the {{domxref('MediaKeySession')}} interface returns a {{jsxref('Promise')}} after generating a license request based on initialization data.
```

Reviewer: You'll also need to update https://developer.mozilla.org/en-US/docs/Web/API/MediaKeySession#mediakeysession.generaterequest.

## Syntax

```js-nolint
generateRequest(initDataType, initData)
```

### Parameters

> - `initDataType`
>   - : A `string` that specifies the type of the `initData`. This parameter is mandatory and cannot be an empty string.

Reviewer: There's no indication here of what this value needs to be, although further down it says the method throws if the data doesn't match the type. So this isn't helpful - noone could use the API, based on this documentation.

It looks like valid values for this are listed in https://www.w3.org/TR/eme-initdata-registry/#registry, so perhaps:

```suggestion
- `initDataType`
  - : A string that specifies the format of the `initData` parameter. This must be one of the following values:
    - `"cenc"`: The `initData` parameter uses the [`"cenc"`](https://www.w3.org/TR/eme-initdata-cenc/) format.
    - `"keyids"`: The `initData` parameter uses the [`"keyids"`](https://www.w3.org/TR/eme-initdata-keyids/) format.
    - `"webm"`: The `initData` parameter uses the [`"webm"`](https://www.w3.org/TR/eme-initdata-webm/) format.
```

Reviewer: btw we never need to say parameters are mandatory. They always are unless we indicate otherwise.

Author: Thank you for letting me know, This is very helpful.

- `initData`

>   - : A `BufferSource` (e.g, {{jsxref("ArrayBuffer")}} , {{jsxref("TypedArray")}} or {{jsxref("DataView")}}) that contains the initialization data. This parameter is mandatory and cannot be an empty array.

Reviewer: ```suggestion
  - : Initialization data for the request, in the format specified by `initDataType`. It is an instance of any of the following types:

    - {{jsxref("ArrayBuffer")}}
    - {{jsxref("DataView")}}
    - {{jsxref("TypedArray")}}
```

### Return value

A {{jsxref('Promise')}}.

### Exceptions

- {{jsxref("TypeError")}}
  - : Raised if `initDataType` is an empty string, if `initData` is an empty array, or if the provided `initData` is not valid according to the specified `initDataType`.
- {{domxref("DOMException")}} `NotSupportedError`

>   - : Raised if the Key System implementation associated with the `MediaKeySession` object does not support the provided `initDataType`, if the sanitized initialization data is empty, or if the sanitized initialization data is not supported by the Content Decryption Module (CDM).

Reviewer: ```suggestion
  - : Raised if the Key System implementation associated with the `MediaKeySession` object does not support the provided `initDataType`, if the sanitized initialization data is empty, or if the sanitized initialization data is not supported by the content decryption module (CDM).
```

We don't capitalize this in https://developer.mozilla.org/en-US/docs/Web/API/MediaKeySession, so let's be consistent. I don't know what "Key System implementation" is and we don't say, however that's probably out of scope for this PR.
